src/search/merge_and_shrink/transition_system.cc

```diff
@@ -328,63 +396,67 @@ void TransitionSystem::apply_abstraction(
-      storing these affected groups). Furthermore, rather than computing
-      locally equivalent labels from scratch, we did not per default add a new
-      group for every label, but checked for an existing equivalent label
-      group. In issue539, it turned out that this incremental fashion of
-      computation does not accelerate the computation.
+      local labels, and the local labels themselves if do not represent any
+      global labels anymore. Also collect the transitions of all reduced
+      labels. Add a new local label for every new label and assign the
+      collected transitions to it. Recompute the cost of all affected local
+      labels and re-compute locally equivalent labels.
     */
 
     if (only_equivalent_labels) {
-        label_equivalence_relation->apply_label_mapping(label_mapping);
+        // Update both label mappings.
+        for (const pair<int, vector<int>> &mapping : label_mapping) {
+            int new_label = mapping.first;
+            const vector<int> &old_labels = mapping.second;
+            int local_label = global_to_local_label[old_labels.front()];
+            LabelGroup &label_group = local_to_global_labels[local_label];
+            label_group.push_back(new_label);
+            global_to_local_label[new_label] = local_label;
+
+            for (int old_label : old_labels) {
+                assert(global_to_local_label[old_label] == local_label);
+                label_group.erase(find(label_group.begin(), label_group.end(), old_label));
+                // Reset (for consistency only, old labels are never accessed).
+                global_to_local_label[old_label] = -1;
+            }
+        }
     } else {
         /*
-          Go over all mappings, collect transitions of old groups and
-          remember all affected group IDs. This needs to happen *before*
-          updating label_equivalence_relation, because after updating it,
-          we cannot find out the group ID of reduced labels anymore.
+          Go over all mappings, collect transitions of old local labels and
+          remember all affected such labels. This needs to happen *before*
+          updating the label mappings (because otherwise, we could not find
+          the correct local label for a given reduced global label).
         */
         vector<vector<Transition>> new_transitions;
         new_transitions.reserve(label_mapping.size());
-        unordered_set<int> affected_group_ids;
+        unordered_set<int> affected_local_labels;
         for (const pair<int, vector<int>> &mapping: label_mapping) {
-            const vector<int> &old_label_nos = mapping.second;
-            assert(old_label_nos.size() >= 2);
-            unordered_set<int> seen_group_ids;
+            const vector<int> &old_labels = mapping.second;
+            assert(old_labels.size() >= 2);
+            unordered_set<int> seen_local_labels;
             set<Transition> new_label_transitions;
```

Comment:
I assume you use set here because you want the transitions ordered. Could it be more efficient to use an unordered_set and sort it once at the end? (I have absolutely no intuition)

Reply:
I also don't have intuition for such questions. I added this as a note to the code but wouldn't want to try it in this issue.

Reply:
FWIW, my intuition is that if you don't have too many many duplicates, it is cheaper to add everything to a vector and then use sort/unique/erase to remove duplicates. This doesn't require the more complex data structures of a set or a hash set. But I also wouldn't change it now.